lib/src/executable.dart

```diff
@@ -57,130 +56,152 @@ Future<int> runExecutable(Entrypoint entrypoint, String package,
-    // directory.
-    packageConfig = p.toUri(p.absolute(packagesFile));
-  }
-
-  var url = p.toUri(p.absolute(path));
-  var argList = args.toList();
-  try {
-    await isolate.runUri(url, argList, null,
+/// Returns the programs's exit code.
+Future<int> _runDartProgram(
+    String path, List<String> args, String packageConfig,
+    {bool enableAsserts, List<String> vmArgs}) async {
+  path = p.absolute(path);
+  packageConfig = p.absolute(packageConfig);
+
+  // We use Isolate.spawnUri when there are no extra vm-options.
+  // That provides better signal handling, and possibly faster startup.
+  if (vmArgs.isEmpty) {
+    var argList = args.toList();
+    await isolate.runUri(p.toUri(path), argList, null,
         enableAsserts: enableAsserts,
         automaticPackageResolution: packageConfig == null,
-        packageConfig: packageConfig);
-  } on IsolateSpawnException catch (error) {
-    if (recompile == null) rethrow;
-    if (!error.message.contains('Invalid kernel binary format version')) {
-      rethrow;
-    }
-
-    log.fine('Precompiled executable is out of date.');
-    await recompile();
-    await isolate.runUri(url, argList, null,
-        enableAsserts: enableAsserts, packageConfig: packageConfig);
+        packageConfig: p.toUri(packageConfig));
+    return exitCode;
+  } else {
+    // By ignoring sigint, only the child process will get it when
+    // they are sent to the current process group. That is what happens when
+    // you send signals from the terminal.
+    //
+    // This allows the child to not be orphaned if it sets up handlers for these
+    // signals.
+    //
+    // We do not drain sighub because it is generally a bad idea to have
+    // non-default handling for it.
+    //
+    // We do not drain sigterm and sigusr1/sigusr2 because it does not seem to
+    // work well in manual tests.
+    //
+    // We do not drain sigquit because dart doesn't support listening to it.
+    // https://github.com/dart-lang/sdk/issues/41961 .
+    //
+    // TODO(sigurdm) To handle signals better we would ideally have `exec`
+    // semantics without `fork` for starting the subprocess.
+    // https://github.com/dart-lang/sdk/issues/41966.
+    unawaited(ProcessSignal.sigint.watch().drain());
+
+    final process = await Process.start(
+      Platform.resolvedExecutable,
+      [
+        '--packages=$packageConfig',
+        ...vmArgs,
+        if (enableAsserts) '--enable-asserts',
+        p.toUri(path).toString(),
+        ...args,
+      ],
+      mode: ProcessStartMode.inheritStdio,
+    );
+
+    return process.exitCode;
   }
+}
+
+/// An executable in a package
+class Executable {
+  String package;
+  // The relative path to the executable inside the root of [package].
+  String relativePath;
+  Executable(this.package, String program)
+      : relativePath = adaptProgramToPath(program);
 
-  return exitCode;
+  static String adaptProgramToPath(String program) {
```

Comment:
This can be private if we call it from the constructor...

but I wanted to avoid entangling this logic in `pub run --dart-dev`. I agree that `pub run` needs to retain this for compatibility. Can we instead of calling it from `Executable` constructor, call it before we pass to `Executable` constructor and then make it public...

Or make it private and give `Executable` two constructors.

Reply:
Let's make two constructors...